lib/event-subscribers.loader.ts

```diff
@@ -31,39 +45,126 @@ export class EventSubscribersLoader
     const providers = this.discoveryService.getProviders();
     const controllers = this.discoveryService.getControllers();
     [...providers, ...controllers]
-      .filter(wrapper => wrapper.isDependencyTreeStatic())
       .filter(wrapper => wrapper.instance)
       .forEach((wrapper: InstanceWrapper) => {
         const { instance } = wrapper;
         const prototype = Object.getPrototypeOf(instance) || {};
+        const isRequestScoped = !wrapper.isDependencyTreeStatic();
         this.metadataScanner.scanFromPrototype(
           instance,
           prototype,
           (methodKey: string) =>
-            this.subscribeToEventIfListener(instance, methodKey),
+            this.subscribeToEventIfListener(
+              instance,
+              methodKey,
+              isRequestScoped,
+              wrapper.host as Module,
```

Comment:
There is no other way of doing this without using `as`? Just to avoid losing type-safety

Reply:
I tried a couple of things but apparently this is the less hacky way of doing it. To give you more context on that, the reason Why I had to do it is because of the following:

1. The `InstanceWrapper.host` is an optional attribute, as you can see [here](https://github.com/nestjs/nest/blob/c2d73f91d7848994302fb739f61c080e9e7f9b20/packages/core/injector/instance-wrapper.ts#L59)

2. This `host` attribute is used by the `injector.loadPerContext` method [in this same file](https://github.com/thiagomini/event-emitter/blob/a2bbb696968b5d95077d8f4a469a2bb8c9ed974f/lib/event-subscribers.loader.ts#L129)

3. `loadPerContext` method requires a `module` attribute, hence it's not optional

4. this project has strict flag set to true, as you can see in the [tsconfig.json](https://github.com/thiagomini/event-emitter/blob/26426c20abc08cf05cb98cb397f625489def805d/tsconfig.json#L5)

The reason **4** is the actual culprit of this `as Module` statement. You can see in the [nestjs/bull package](https://github.com/nestjs/bull/blob/87aba5742d40853d50b4a10dc4d63f4e7352ee39/packages/bull/lib/bull.explorer.ts#L81) the module is handed over the function with no casting, and that's because its [tsconfig](https://github.dev/nestjs/bull/blob/f16bf2b31e5c54af73e3f9a2e1619a21ab122970/packages/tsconfig.build.json#L15) file does not have `strict: true`

Reply:
if it's optional, can't you just `if` it?

Reply:
You mean to add an `if` statement before calling `subscribeToEventIfListener` ?

Reply:
Yes.
Maybe like 
```ts
if(!wrapper.host) {
  throw new Error()
}
```

Reply:
Hmmmm I see, but do you think this is really the expected behavior here? To be honest, I'm not sure what would happen if this host really didn't exist when we tried to inject the instance. @kamilmysliwiec  do you have some opinion on that as wel?

Reply:
At that time, `host` will never be `undefined` so the `as Module` assertion is correct here.